src/passes/Stackify.cpp

```diff
@@ -0,0 +1,497 @@
+  switch (scope.kind) {
+    case Scope::Block:
+      patchInstrs(curr);
+      break;
+    case Scope::Loop:
+      patchInstrs(curr->cast<Loop>()->body);
+      break;
+    case Scope::If:
+      patchInstrs(curr->cast<If>()->ifTrue);
+      break;
+    case Scope::Else:
+      patchInstrs(curr->cast<If>()->ifFalse);
+      break;
+    case Scope::Catch:
+      patchInstrs(curr->cast<Try>()->catchBodies.back());
+      break;
+    case Scope::Try:
+      WASM_UNREACHABLE("try without catch");
+    case Scope::Func:
+      WASM_UNREACHABLE("unexpected end of function");
+  }
+  scopeStack.pop_back();
+  scopeStack.back().instrs.push_back(curr);
+}
+
+void Stackifier::emitFunctionEnd() {
+  auto& scope = scopeStack.back();
+  assert(scope.kind == Scope::Func);
+  patchInstrs(func->body);
+}
+
+void Stackifier::emitUnreachable() {
+  // TODO: Try making this a nop
+  auto& instrs = scopeStack.back().instrs;
+  instrs.push_back(builder.makeUnreachable());
+}
+
+void Stackifier::emitTupleExtract(TupleExtract* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  auto types = curr->tuple->type;
+  // Drop all the values after the one we want
+  for (size_t i = types.size() - 1; i > curr->index; --i) {
+    instrs.push_back(builder.makeDrop(builder.makePop(types[i])));
+  }
+  // If the extracted value is the only one left, we're done
+  if (curr->index == 0) {
+    return;
+  }
+  // Otherwise, save it to a scratch local and drop the other values
+  auto type = types[curr->index];
+  Index scratch = getScratchLocal(type);
+  instrs.push_back(builder.makeLocalSet(scratch, builder.makePop(type)));
+  for (size_t i = curr->index - 1; i != size_t(-1); --i) {
+    instrs.push_back(builder.makeDrop(builder.makePop(types[i])));
+  }
+  // Retrieve the saved value
+  instrs.push_back(builder.makeLocalGet(scratch, type));
+}
+
+void Stackifier::emitDrop(Drop* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    // Drop each element individually
+    auto types = curr->value->type;
+    for (auto it = types.rbegin(), end = types.rend(); it != end; ++it) {
+      instrs.push_back(builder.makeDrop(builder.makePop(*it)));
+    }
+  } else {
+    poppify(curr);
+    instrs.push_back(curr);
+  }
+}
+
+void Stackifier::emitLocalGet(LocalGet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->type.isTuple()) {
+    auto types = func->getLocalType(curr->index);
+    const auto& elems = tupleVars[curr->index];
+    for (size_t i = 0; i < types.size(); ++i) {
+      instrs.push_back(builder.makeLocalGet(elems[i], types[i]));
+    }
+  } else {
+    instrs.push_back(curr);
+  }
+}
+
+void Stackifier::emitLocalSet(LocalSet* curr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (curr->value->type.isTuple()) {
+    auto types = func->getLocalType(curr->index);
+    const auto& elems = tupleVars[curr->index];
+    // Add the unconditional sets
+    for (size_t i = types.size() - 1; i >= 1; --i) {
```

Comment:
I'm confused as to why `0` is special here. It seems like for tee you emit
```
set 3
set 2
set 1
tee 0
get 1
get 2
get 3
```
Why is that correct?

Reply:
Oh, wait, I see it, nevermind...